### PR TITLE
refactor(binder): in-place `cast_mut` or `ExprImpl::take`

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR contains 2 candidate solutions to simplify the repeated pattern below. **We only need one of them.**
```
let expr_owned = std::mem::replace(expr_mut_ref, new_dummy_expr());
*expr_mut_ref = expr_owned.cast_implicit(target_data_type)?;
```

Two straightforward but less desired approaches are:
* Extract this into a reusable `cast_mut` utility function
* `impl Default for ExprImpl` so that we can just `take` rather than `replace` with a dummy value

The first is not a proper abstraction because it may leave the input partially mutated. When the cast fails, the input is already replaced by the dummy expression. It is easier for the caller to have input unchanged when the function call fails.

As for the second, it is hard to select a proper *default* expression that is universally meaningful beyond the use case here. In this case anything works because it is immediately overwritten (or unused when the cast error bubbles up to the user).

The implementations in this PR are slightly improved over the 2 ideas above:
* Define the owned version `cast_implicit` as wrapper of `cast_mut`, rather than `cast_mut` over `cast_implicit`. This way we have finer control to only mutate at last after check for all possible errors (except for `try_for_each`, which may worth further imporvement).
* Define `ExprImpl::take` to update the expression to `null` of its original data type. This way the expression remains properly typed and it is easily accepted that `take` results in `null`.

Most of this PR is about the first approach, which refactors the existing `new_cast` from taking owned `child: ExprImpl` to `child: &mut ExprImpl`. The second is only used at one place that is not related cast - an expr rewriter in BoundJoin.

We only need to choose one of them. The first requires more code change, results in wired naming and signature (`new_cast(&mut ExprImpl) -> Result<()>`), and covers only cast but not expr rewriter. The second feels a little bit like cheating because we are actually cloning its return type - a more naive way of getting owned value from a mut ref is just to clone everything

Your opinions are welcomed.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

### CODE REVIEW

File: src/frontend/src/expr/function_call.rs [hunk 1]
```diff
@@ -111,19 +111,19 @@ impl FunctionCall {
 
     /// Create a cast expr over `child` to `target` type in `allows` context.
```
Comment (BugenZhao) [inline src/frontend/src/expr/function_call.rs:112]: Should we note somewhere that developer need to ensure not to touch `child` if to return an error?

File: src/frontend/src/expr/function_call.rs [hunk 2]
```diff
@@ -111,19 +111,19 @@ impl FunctionCall {
...
     pub fn new_cast(
```
Comment (BugenZhao) [inline src/frontend/src/expr/function_call.rs:113]: What about renaming this function as  something like `cast_mut`?

**Review by BugenZhao — APPROVED** [review]
LGTM

**Review by st1page — APPROVED** [review]

**Comment by xxchan** [comment]
> LGTM

So which one LGTY? 🤣

**Comment by xiangjinwu** [comment]
I am going to keep both, at least for some time until we find one of them is problematic.

The refactor of `cast` does make it more general, as functions have owned `ExprImpl` can always call a `&mut ExprImpl` function easily.

The `take` has a clear semantic and is lighter than a full `clone`.

**Pull request merged**